xarray/core/datatree.py

```diff
@@ -438,6 +438,7 @@ class DataTree(
     _cache: dict[str, Any]  # used by _CachedAccessor
     _data_variables: dict[Hashable, Variable]
     _node_coord_variables: dict[Hashable, Variable]
+    _node_indexed_coord_variables: dict[Hashable, Variable]
```

Comment:
If these are indexed, should the type here be `IndexVariable`?

Reply:
Or should we have just `_node_coord_variables` and which ones are indexed is determined by examining `_node_indexes`? (So `_node_indexed_coord_variables becomes a derived property)

Reply:
Or should we have 
```python
_node_non_indexed_coord_variables
_node_indexed_coord_variables
_node_indexes
```
The current naming is kind of unclear whether `_node_coord_variables` includes `_node_indexed_coord_variables` or not.

Reply:
My main thinking with this implementation was to make accessing the `_coord_variables` property linear in the depth of the tree, not linear in the number of coordinate variables (I'll add a comment about this in the code).

I'll play around with an alternative version here that avoids the need to store the extra state variable.